sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnInvokers.java

```diff
@@ -0,0 +1,500 @@
+            .withParameter(fnClass)
+            .intercept(new InvokerConstructor())
+            // Delegate processElement(), startBundle() and finishBundle() to the fn.
+            .method(ElementMatchers.named("invokeProcessElement"))
+            .intercept(new ProcessElementDelegation(signature.getProcessElement()))
+            .method(ElementMatchers.named("invokeStartBundle"))
+            .intercept(
+                signature.getStartBundle() == null
+                    ? new NoopMethodImplementation()
+                    : new BundleMethodDelegation(signature.getStartBundle()))
+            .method(ElementMatchers.named("invokeFinishBundle"))
+            .intercept(
+                signature.getFinishBundle() == null
+                    ? new NoopMethodImplementation()
+                    : new BundleMethodDelegation(signature.getFinishBundle()))
+            .method(ElementMatchers.named("invokeSetup"))
+            .intercept(
+                signature.getSetup() == null
+                    ? new NoopMethodImplementation()
+                    : new LifecycleMethodDelegation(signature.getSetup()))
+            .method(ElementMatchers.named("invokeTeardown"))
+            .intercept(
+                signature.getTeardown() == null
+                    ? new NoopMethodImplementation()
+                    : new LifecycleMethodDelegation(signature.getTeardown()));
+
+    DynamicType.Unloaded<?> unloaded = builder.make();
+
+    @SuppressWarnings("unchecked")
+    Class<? extends DoFnInvoker<?, ?>> res =
+        (Class<? extends DoFnInvoker<?, ?>>)
+            unloaded
+                .load(DoFnInvokers.class.getClassLoader(), ClassLoadingStrategy.Default.INJECTION)
+                .getLoaded();
+    return res;
+  }
+
+  /** Implements an invoker method by doing nothing and immediately returning void. */
+  private static class NoopMethodImplementation implements Implementation {
+    @Override
+    public InstrumentedType prepare(InstrumentedType instrumentedType) {
+      return instrumentedType;
+    }
+
+    @Override
+    public ByteCodeAppender appender(final Target implementationTarget) {
+      return new ByteCodeAppender() {
+        @Override
+        public Size apply(
+            MethodVisitor methodVisitor,
+            Context implementationContext,
+            MethodDescription instrumentedMethod) {
+          StackManipulation manipulation = MethodReturn.VOID;
+          StackManipulation.Size size = manipulation.apply(methodVisitor, implementationContext);
+          return new Size(size.getMaximalSize(), instrumentedMethod.getStackSize());
+        }
+      };
+    }
+  }
+
+  /**
+   * Base class for implementing an invoker method by delegating to a method of the target {@link
+   * DoFn}.
+   */
+  private abstract static class MethodDelegation implements Implementation {
+    FieldDescription delegateField;
+
+    @Override
+    public InstrumentedType prepare(InstrumentedType instrumentedType) {
```

Comment:
It might be better to pass the `delegateField` in to the constructor and mark it final.